### PR TITLE
Show/Hide filterTab based on Data

### LINKED ISSUE #552: App Bombs when the ID Column in the Filter Domain isn't specified
Need to disable the filterDomain until the id_column is defined. Maybe Just use some if/else logic with `req()`?

### LINKED ISSUE #555: Fix broken sections of filter page

### PR DESCRIPTION
**Pull request by jwildfire**
# Overview

Filter tab should be deactivated when the data and metadata don't support filtering. This PR implements a series of checks in the new `mod_filterTabChecks.R` function, which are then used in `mod_filterTab` to determine whether the page should be created. 

For example, if `id_col` is not defined in the metadata for the specified `filterDomain`, then it's not possible to filter the data across all domains. However, when `id_col` is set on the data/mapping tab, the filter tab should be activated.

Fixes #552. Fix #555. 

# Test Notes

Easiest way to test is to run the app at `tests/testthat/module_examples/filterTab/app.R`, and checking the associated `testthat` tests.  Also might be worth trying out some of the custom builds documented in this [wiki page](https://github.com/SafetyGraphics/safetyGraphics/wiki/Vignette:-Shiny-App-Customization-Cookbook) with a focus on whether the filterTab is found.

### CODE REVIEW

**Comment by jwildfire** [comment]
@samussiah Filed a few issues to deal with later, but I think this is ok to review/merge now. Let me know if you have any questions.

File: R/app_startup.R [hunk 1]
```diff
@@ -42,12 +50,19 @@ app_startup<-function(domainData=NULL, meta=NULL, charts=NULL, mapping=NULL, cha
         mapping<-bind_rows(mapping_list, .id = "domain")
     }
 
+    # Set filterDomain to null if no data exists
+    if(!filterDomain %in% names(domainData)){
```
Comment (samussiah) [inline R/app_startup.R:54]: Only issue with this is if `filterDomain == NULL` this check bombs.  Obviously that's never going to actually happen outside of testing but worth noting.

File: R/mod_filterTabChecks.R [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                id_col <- reactive({
+                    filter_data <- current_mapping() %>% filter(.data$domain==filterDomain)   
+                    id<- filter_data %>% filter(.data$text_key=="id_col")%>%pull(.data$current)
+                    return(id)
+                })
+
+                id_check <- all(domainData %>% purrr::map_lgl(~{id_col() %in% colnames(.x)}))
+                if(!id_check){
+                    currentStatus <- FALSE
+                    filterCheckNote <- "id_col is not found in one or more data domain. Update mappings and try again. "
```
Comment (samussiah) [inline R/mod_filterTabChecks.R:47]: Works as expected.

File: R/mod_filterTabChecks.R [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    # Check to see if data can be filtered using current settings.
+    filterCheck<-reactive({
+        currentStatus <- TRUE
+        filterCheckNote<-NULL
+        if(is.null(filterDomain)){
+            # Make sure filterDomain exists.
+            currentStatus <- FALSE
+            filterCheckNote <- "the filter domain is not specified."
+        }else if(!(filterDomain %in% names(domainData))){
```
Comment (samussiah) [inline R/mod_filterTabChecks.R:22]: Works as expected.

File: R/mod_filterTabChecks.R [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            currentStatus <- FALSE
+            filterCheckNote <- "the specified filter domain is not found in provided data."
+        }else{
+            # Make sure id_col is specified for filterDomain in the mapping and has 1+ characters          
+            filter_data <- current_mapping() %>% 
+                filter(.data$domain==filterDomain)  %>%
+                filter(.data$text_key=="id_col")          
+            id_col_found <- nrow(filter_data)==1 & nchar(filter_data[1,"current"])>0
+            
+            if(!id_col_found){
```
Comment (samussiah) [inline R/mod_filterTabChecks.R:33]: Works as expected.

File: R/mod_filterTabChecks.R [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                })
+
+                id_check <- all(domainData %>% purrr::map_lgl(~{id_col() %in% colnames(.x)}))
+                if(!id_check){
+                    currentStatus <- FALSE
+                    filterCheckNote <- "id_col is not found in one or more data domain. Update mappings and try again. "
+                }else if(FALSE){
+                    # TODO: Warn, but proceed if id_col is non-unique in filter domain.
+                }else if(FALSE){
+                    # TODO: Warn, but proceed if id_col differs across domains.
```
Comment (samussiah) [inline R/mod_filterTabChecks.R:51]: This edge case with different ID names across domains doesn't work but I assumed `id_col()` only ever resolves to a single value... although looking at line 38 above `id_col()` could resolve differently depending on the domain.

File: tests/testthat/module_examples/filterTab/app.R [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+            ),
+            collapse="/"
+        )
+    )
+
+    ###################
+    #### Example 4 ####
+    ###################
+    mapping4<-ex1_mapping
+    mapping4[mapping4$domain=="dm" & mapping4$text_key=="id_col","text_key"]<-"Not_id_col"
```
Comment (samussiah) [inline tests/testthat/module_examples/filterTab/app.R:147]: I'm wondering when one would edit the meta.  Anyway, A+ logic.

File: tests/testthat/module_examples/filterTab/app.R [hunk 7]
```diff
@@ -0,0 +1,205 @@
+library(shiny)
+library(dplyr)
+library(shinydashboard)
+library(safetyGraphics)
+
+#reactlogReset()
+
+ui <- tagList(
```
Comment (samussiah) [inline tests/testthat/module_examples/filterTab/app.R:8]: testing app, cooolllll

File: tests/testthat/test_mod_filterTab.R [hunk 8]
```diff
@@ -0,0 +1,51 @@
+context("Tests for the filterTab R module")
+library(safetyGraphics)
+library(shiny)
+library(shinytest)
+library(testthat)
+library(stringr)
+
+app <- ShinyDriver$new("./module_examples/filterTab")
```
Comment (samussiah) [inline tests/testthat/test_mod_filterTab.R:8]: so the app above is called here...

File: tests/testthat/test_mod_filterTab.R [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+initial<-app$getAllValues()
```
Comment (samussiah) [inline tests/testthat/test_mod_filterTab.R:9]: `getAllValues` returns... calls to `exportTestValues`?

**Review by samussiah — APPROVED** [review]
Found a couple of edge cases, one of which you noted in a `TODO`, that would likely only turn up in testing so consider this PR approved.  Nice logic Spock.

File: R/app_startup.R (hunk 1 above)
Comment (jwildfire) [reply R/app_startup.R:54]: Hm, I think this is actually worth fixing. I don't think it would be crazy for someone to call `safetyGraphicsApp(filterDomain=NULL)` to initialize the app with no filtering. And that does indeed bomb on this line. 

Good catch!

File: R/mod_filterTabChecks.R (hunk 5 above)
Comment (jwildfire) [reply R/mod_filterTabChecks.R:51]: Yeah - Only the `id_col()` for `filter_domain()` is used here. I thought about creating a single `id_col` control that cut across all domains, but decided in the end that it adds a bit more complexity than I wanted to deal with (at least for v2.0). 

I think this logic will work ok for now - although it could definitely lead to some confusing results in certain situations. where id_col is different in different domains or where it is non-unique.

File: tests/testthat/module_examples/filterTab/app.R (hunk 6 above)
Comment (jwildfire) [reply tests/testthat/module_examples/filterTab/app.R:147]: Yeah, I think custom meta files will happen occasionally, but a bit edge case-y for now.

File: tests/testthat/test_mod_filterTab.R (hunk 9 above)
Comment (jwildfire) [reply tests/testthat/test_mod_filterTab.R:9]: It returns a snap shot of all input, output and test values. Full docs here: https://rstudio.github.io/shinytest/reference/ShinyDriver.html#method-getAllValues

File: R/mod_filterTabChecks.R (hunk 5 above)
Comment (jwildfire) [reply R/mod_filterTabChecks.R:51]: OK, done in 079d579e87581c823fc29e8f46ffa75de06dca81
